src/portalnet/socket.rs

```diff
@@ -0,0 +1,56 @@
+use log::debug;
+use std::collections::VecDeque;
+use std::io;
+use std::net::{SocketAddr, UdpSocket};
+
+use get_if_addrs::{get_if_addrs, IfAddr, Interface};
+
+/// Ping a STUN server on the public network. This does two things:
+/// - Creates an externally-addressable UDP port, if you are behind a NAT
+/// - Returns the public IP and port that corresponds to your local port
+pub fn stun_for_external(local_socket_addr: &SocketAddr) -> Option<SocketAddr> {
+    let socket = UdpSocket::bind(local_socket_addr).unwrap();
+    let external_addr = stunclient::StunClient::new("143.198.142.185:3478".parse().unwrap())
+        .query_external_address(&socket);
+    debug!(
+        "STUN claims that public network endpoint is: {:?}",
+        external_addr,
+    );
+    external_addr.ok()
+}
+
+pub fn default_local_address(port: u16) -> SocketAddr {
+    let ip = match first_interface() {
+        Some(interface) => interface.addr.ip(),
+        None => "127.0.0.1".parse().unwrap(),
+    };
```

Comment:
[`map_or`](https://doc.rust-lang.org/std/option/enum.Option.html#method.map_or) or `map_or_else` might be neater